File: BEmu3/MarketDataRequest/EventMarket.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BEmu3.MarketDataRequest
{
    public class EventMarket : Event
    {
        private readonly List<Message> _messages;

        internal EventMarket(EventType evtType, CorrelationID corr, IEnumerable<Subscription> subscriptions) //use for subscribing
        {
            this._messages = new List<Message>();

            switch (evtType)
            {
                case EventType.SESSION_STATUS:
                    base._type = evtType;
                    MessageMarketSessionOpened msgSessionOpened = new MessageMarketSessionOpened();
                    this._messages.Add(msgSessionOpened);
                    break;

                case EventType.SERVICE_STATUS:
                    base._type = evtType;
                    MessageMarketServiceStatus msgServiceStatus = new MessageMarketServiceStatus(corr);
                    this._messages.Add(msgServiceStatus);
                    break;

                case EventType.SUBSCRIPTION_STATUS:
                    base._type = evtType;
                    foreach (var item in subscriptions)
                    {
                        MessageMarketSubscriptionStarted msgSubStatus = new MessageMarketSubscriptionStarted(item);
                        this._messages.Add(msgSubStatus);
                    }
                    break;

                case EventType.SUBSCRIPTION_DATA:
                    base._type = evtType;
                    foreach (var item in subscriptions)
                    {
                        MessageMarketSubscriptionData msgSubData = new MessageMarketSubscriptionData(item, EventMarket.GenerateFakeMessageData(item));
                        this._messages.Add(msgSubData);
                    }
                    break;

                default:
                    throw new NotImplementedException(string.Format("BEmu3.MarketDataRequest.EventMarket.EventMarket: doesn't expect EventType {1}", evtType.ToString()));
            }
        }

        internal EventMarket(EventType evtType, Subscription subscription) //use for unsubscribing
        {
            this._messages = new List<Message>();

            switch (evtType)
            {
                case EventType.SUBSCRIPTION_STATUS:
                    base._type = evtType;
                    MessageMarketSubscriptionCanceled msgCancel = new MessageMarketSubscriptionCanceled(subscription);
                    this._messages.Add(msgCancel);
                    break;
                default:
                    break;
            }
        }

        private static readonly Random _random = new Random(100);
        private static Dictionary<string, object> GenerateFakeMessageData(Subscription sub)
        {
            var result = new Dictionary<string, object>();
            foreach (var item in sub.Fields)
            {
                string upper = item.ToUpper();

                if (upper.Contains("TIME"))
                    result.Add(item, new Types.TimeType(DateTime.Now.AddMinutes(-1d * EventMarket._random.NextDouble() * 100d)));
                else if (upper.Contains("DATE"))
                    result.Add(item, new Types.DateType(DateTime.Now.AddDays(-1d * EventMarket._random.NextDouble() * 100d)));
                else
                    result.Add(item, Math.Round(EventMarket._random.NextDouble() * 100d, 2));
            }
            return result;
        }

        public override IEnumerable<Message> GetMessages()
        {
            return this._messages;
        }
    }
}

File: BEmu_cpp/src/MarketDataRequest/EventMarket.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/MarketDataRequest/EventMarket.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "MarketDataRequest/EventMarket.h"
#include "MarketDataRequest/MessageMarketSessionOpened.h"
#include "MarketDataRequest/MessageMarketServiceStatus.h"
#include "MarketDataRequest/MessageMarketSubscriptionFailure.h"
#include "MarketDataRequest/MessageMarketSubscriptionStarted.h"
#include "MarketDataRequest/MessageMarketSubscriptionData.h"
#include "MarketDataRequest/MessageMarketSubscriptionCanceled.h"

#include "Types/Rules.h"
#include "Types/RandomDataGenerator.h"

namespace BEmu
{
	namespace MarketDataRequest
	{
		EventMarket::EventMarket(Event::EventType evtType, CorrelationId corr, SubscriptionList subs) : EventPtr(0)
		{
			this->_messages = new std::vector<MessagePtr*>();

			switch(evtType)
			{
				case Event::SESSION_STATUS:
				{
					this->_type = evtType;
					MessageMarketSessionOpened * msgSessionOpened = new MessageMarketSessionOpened();
					this->_messages->push_back(msgSessionOpened);
					break;
				}

				case Event::SERVICE_STATUS:
				{
					this->_type = evtType;
					MessageMarketServiceStatus * msgServiceStatus = new MessageMarketServiceStatus(corr);
					this->_messages->push_back(msgServiceStatus);
					break;
				}

				case Event::SUBSCRIPTION_STATUS:
				{
					this->_type = evtType;
					std::vector<Subscription> * list = subs.list();
					for(std::vector<Subscription>::const_iterator iter = list->begin(); iter != list->end(); ++iter)
					{
						Subscription sub = *iter;
						bool securityError = Rules::IsSecurityError(sub.security());
						if(securityError)
						{
							MessageMarketSubscriptionFailure * msgError = new MessageMarketSubscriptionFailure(sub);
							this->_messages->push_back(msgError);
						}
						else
						{
							MessageMarketSubscriptionStarted * msgSubStatus = new MessageMarketSubscriptionStarted(sub);
							this->_messages->push_back(msgSubStatus);
						}
					}
					break;
				}

				case Event::SUBSCRIPTION_DATA:
				{
					this->_type = evtType;
					for(std::vector<Subscription>::const_iterator iter = subs.list()->begin(); iter != subs.list()->end(); ++iter)
					{
						Subscription sub = *iter;
						bool securityError = Rules::IsSecurityError(sub.security());
						if (!securityError)
						{
							MessageMarketSubscriptionData * msgSubData = new MessageMarketSubscriptionData(sub, EventMarket::generateFakeMessageData(sub));
							this->_messages->push_back(msgSubData);
						}
					}
					break;
				}

				default:
					throw eventPtrEx;
			}
		}

		EventMarket::EventMarket(Event::EventType evtType, Subscription sub) : EventPtr(0)
		{
			this->_messages = new std::vector<MessagePtr*>();

			switch(evtType)
			{
				case Event::SUBSCRIPTION_STATUS:
				{
					this->_type = evtType;
					MessageMarketSubscriptionCanceled * msgCancel = new MessageMarketSubscriptionCanceled(sub);
					this->_messages->push_back(msgCancel);
				}
				break;
			}
		}

		EventMarket::~EventMarket()
		{
			delete this->_messages;
		}

		std::vector<MessagePtr*> * EventMarket::getMessages() const
		{
			return this->_messages;
		}

		std::map<std::string, ObjectType> EventMarket::generateFakeMessageData(Subscription sub)
		{
			return RandomDataGenerator::GetMarketDataFields(sub.fieldList());
		}

	}
}